### PR TITLE
Experimental simplifier flag on main + use the flag with 2 tests

### PR DESCRIPTION
**Pull request by assem2002**

### CODE REVIEW

File: tests/reference/asr-template_04-f41dd3e.stdout [hunk 1]
```diff
@@ -508,8 +508,62 @@
                                                     ((Var 198 tmp))]
                                                     (Array
                                                         (Integer 4)
-                                                        [(()
-                                                        ())]
+                                                        [((IntegerConstant 1 (Integer 4) Decimal)
```
Comment (assem2002) [inline tests/reference/asr-template_04-f41dd3e.stdout:511]: I've ported minimal amount of code to get simplifier.cpp to compile. Is that nomral having these huge ASR changes?

File: integration_tests/CMakeLists.txt [hunk 2]
```diff
@@ -261,8 +261,8 @@ RUN(NAME stop_03 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME print_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc cpp llvm2 wasm fortran)
 RUN(NAME print_02 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 wasm fortran)
 RUN(NAME print_03 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 wasm c fortran)
-RUN(NAME print_04 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 fortran)
-RUN(NAME print_05 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 fortran)
+RUN(NAME print_04 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 fortran EXTRA_ARGS --experimental-simplifier)
+RUN(NAME print_05 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvm2 fortran EXTRA_ARGS --experimental-simplifier)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:265]: I would add a label, I think the extra args are always passed in.

Comment (assem2002) [reply integration_tests/CMakeLists.txt:265]: Yeah that's how we should do things. I was just testing things out.
So every test would use the label but the ones that are still in need of the complete changes from simplifier branch, right?

**Comment by kmr-srbh** [comment]
@assem2002 if this PR is ready, let's merge it with the changes related to the `--experimental-simplifier` option introduction. This will help us start porting fixes into `main`. Let's introduce the `simplifier_pass` label for CI in a subsequent PR.

**Comment by kmr-srbh** [comment]
Okay, so it seems it's trivial to add the `simplifier` label. Let's add it here and merge this PR to get started.

```diff
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ git diff
diff --git a/integration_tests/CMakeLists.txt b/integration_tests/CMakeLists.txt
index 5d3973a0a..fa80d06c2 100644
--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -151,6 +151,10 @@ macro(RUN_UTIL RUN_FAIL RUN_NAME RUN_FILE_NAME RUN_LABELS RUN_EXTRAFILES RUN_EXT
         elseif (LFORTRAN_BACKEND STREQUAL "gfortranModfileCompilation")
             execute_process(COMMAND gfortran ${CMAKE_CURRENT_SOURCE_DIR}/${file_name}.f90 -c)
             add_test(${name} ${CURRENT_BINARY_DIR}/${name})
+        elseif (LFORTRAN_BACKEND STREQUAL "simplifier")
+            add_executable(${name} ${file_name}.f90 ${extra_files})
+            target_compile_options(${name} PUBLIC --experimental-simplifier ${extra_args})
+            add_test(${name} ${CURRENT_BINARY_DIR}/${name})
         else ()
             add_executable(${name} ${file_name}.f90 ${extra_files})
             target_compile_options(${name} PUBLIC ${gfortran_args})
@@ -180,7 +184,7 @@ macro(RUN)
                         "${multiValueArgs}" ${ARGN} )
 
     foreach(b ${RUN_LABELS})
-        if (NOT (b MATCHES "^(llvm|llvm2|llvm_rtlib|c|cpp|x86|wasm|gfortran|llvmImplicit|llvmStackArray|fortran|c_nopragma|llvm_nopragma|llvm_wasm|llvm_wasm_emcc|llvm_omp|mlir|nollvm17orhigher)$"))
+        if (NOT (b MATCHES "^(llvm|llvm2|llvm_rtlib|c|cpp|x86|wasm|gfortran|llvmImplicit|llvmStackArray|fortran|c_nopragma|llvm_nopragma|llvm_wasm|llvm_wasm_emcc|llvm_omp|mlir|nollvm17orhigher|simplifier)$"))
             message(FATAL_ERROR "Unsupported backend: ${b}")
         endif()
     endforeach()
```

**Comment by assem2002** [comment]
Yeah we need to merge this so we can start porting all the changes from the simplifier branch to main branch.
I think we should introduce a label called `NO-EXPERIMENTAL-SIMPLIFIER` as lots of test work with simplifier branch, just few face some issue (compared to the number of working tests).
I'll get the label working and push the changes

**Comment by kmr-srbh** [comment]
Ah, I think we want to test both with and without the `simplifier` pass. https://lfortran.zulipchat.com/#narrow/channel/452774-Internal/topic/Simplifier.20pass/near/477724371

**Comment by assem2002** [comment]
Yeah I think that's what Ondrej wanted, but I'm getting mixed signals lately about what we will do with the simplifier pass.
If this is the case, so let's use your suggestions here https://github.com/lfortran/lfortran/pull/5095#issuecomment-2424702890

**Comment by assem2002** [comment]
I'm wrapping it up to make sure it passes CI. how we'll go from here is to start porting changes from simplifier branch to main and add the label `experimetnal-simplifier` with every time we make port chunks from simplifier branch into main, right?

**Comment by kmr-srbh** [comment]
I see that we are adding the complete file `simplifier.cpp` here in one go. Can we comment parts of it and uncomment on a case by case basis as we go on enabling tests? 

CC @certik

File: integration_tests/CMakeLists.txt (hunk 2 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:265]: Yes. Alternatively we can use the "no" approach and only label tests that don't work. Whatever is easier.

File: src/libasr/asdl_cpp.py [hunk 3]
```diff
@@ -1504,10 +1520,15 @@ def visitField(self, field):
...
-                        self.emit("current_expr = current_expr_copy_%d;" % (self.current_expr_copy_variable_count), level)
+                        one_or_zero = field.name == "value"
+                        if field.name == "value" or field.name == "symbolic_value":
+                            self.emit("if (call_replacer_on_value) {", level)
+                        self.emit("ASR::expr_t** current_expr_copy_%d = current_expr;" % (self.current_expr_copy_variable_count), level + one_or_zero)
+                        self.emit("current_expr = &(x->m_%s);" % (field.name), level + one_or_zero)
+                        self.emit("self().replace_%s(x->m_%s);" % (field.type, field.name), level + one_or_zero)
+                        self.emit("current_expr = current_expr_copy_%d;" % (self.current_expr_copy_variable_count), level + one_or_zero)
+                        if field.name == "value" or field.name == "symbolic_value":
+                            self.emit("}", level)
```
Comment (certik) [inline src/libasr/asdl_cpp.py:1531]: All the changes here might have regressions in them --- do we need it for the initial PR? If not, I would do this later as a dedicated PR just for this change, and we need to carefully test it.

**Review by certik — COMMENTED** [review]
There are changes in template tests from DescriptorArray to FixedSizeArray. I am not sure if they are correct, and even if they are, this should be a separate PR.

The goal of this PR is to introduce the experimental flag, enable it for some tests that work, and ensure there are NO changes in existing tests, guaranteeing no possible regressions.

I would not do changes in `asdl_cpp`, unless they are needed for the simplifier? If they are needed, then keep them, but we need to carefully ensure there are no regressions.

**Comment by certik** [comment]
> Can we comment parts of it and uncomment on a case by case basis as we go on enabling tests? I think we are adding all the changes here.

Yes, I would comment things out, so that we can make this PR not modify any tests, and not modify `asdl_cpp`, or do any other changes that can have regressions in them. We will port all changes later, on a case by cases basis, and ensure no regressions.

File: src/libasr/asdl_cpp.py (hunk 3 above)
Comment (assem2002) [reply src/libasr/asdl_cpp.py:1531]: I ported some changes from simplifier branch as some of them are needed by `simplifier.cpp`. but I'll just try to port the minimal code possible to both make `simplifier.cpp` work + avoid code that might cause regression.

Comment (certik) [reply src/libasr/asdl_cpp.py:1531]: Yes, comment out parts inside `simplifier.cpp` as needed, just to get things to compile. We'll then start sending lots of small separate PRs to port individual parts.

File: integration_tests/CMakeLists.txt [hunk 4]
```diff
@@ -134,7 +136,7 @@ macro(RUN_UTIL RUN_FAIL RUN_NAME RUN_FILE_NAME RUN_LABELS RUN_EXTRAFILES RUN_EXT
             add_test(${name} ${CURRENT_BINARY_DIR}/${name})
         elseif (LFORTRAN_BACKEND STREQUAL "llvm")
             add_executable(${name} ${file_name}.f90 ${extra_files})
-            target_compile_options(${name} PUBLIC ${extra_args})
+            target_compile_options(${name} PUBLIC $<$<COMPILE_LANGUAGE:Fortran>:${extra_args}>)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:139]: What is this change doing?

Comment (assem2002) [reply integration_tests/CMakeLists.txt:139]: For binc tests like `lfortran main.f90 bindc.c` if we add `--experimental-simpifier` to `extra_args` variable, `target_compile_options` would end up with these commands `lfortran main.f90 --experimental-simplifier` and `gcc bindc.c --experimental-simplifier` which of course raises compile error from gcc side.
So we here just specify the compiler options only to get passed while comliling fortran files.

**Comment by assem2002** [comment]
I just mimced  how we deal with `--fast` flag tests

File: integration_tests/CMakeLists.txt (hunk 4 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:139]: I see. How is it that the same exact issue does not happen with `--fast`?

The reason I am asking is to ensure we are not introducing a bug somewhere.

**Comment by certik** [comment]
Nice, just one failure for LLVM >= 17:
```
LCOMPILERS_ASSERT failed: /home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp
function CreateInBoundsGEP(), line number 1748 at 
type
```
The other tests already pass. It seems a little bit too good to be true that just a few tests have to be skipped, but if this works, then this will be easier than I thought.

File: integration_tests/run_tests.py [hunk 5]
```diff
@@ -82,6 +82,8 @@ def get_args():
                 help="Run supported tests with --fast")
     parser.add_argument("-m", action='store_true',
                 help="Check that all module names are unique")
+    parser.add_argument("-experimental-simplifier", "---experimental_simplifier",
```
Comment (certik) [inline integration_tests/run_tests.py:85]: Not sure what is going on here, but for longer options we should just use two dashes `--`:
```suggestion
    parser.add_argument("--experimental-simplifier",
```

File: integration_tests/CMakeLists.txt (hunk 4 above)
Comment (assem2002) [reply integration_tests/CMakeLists.txt:139]: All test that have bindc have `NOFAST` option
for example 
`RUN(NAME lfortran_intrinsic_sin LABELS gfortran llvm EXTRAFILES lfortran_intrinsic_sin_c.c NOFAST)`

File: integration_tests/run_tests.py (hunk 5 above)
Comment (assem2002) [reply integration_tests/run_tests.py:85]: yeah that's right. It was a typo from me.
so the convention is `-` for single character `--` for words or long flag names.

**Comment by assem2002** [comment]
The pass doesn't do any modifications on the ASR (removed all visitor & replacers). I just kept helper functions.

> Nice, just one failure for LLVM >= 17:
> 
> ```
> LCOMPILERS_ASSERT failed: /home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp
> function CreateInBoundsGEP(), line number 1748 at 
> type
> ```
> 
> The other tests already pass. It seems a little bit too good to be true that just a few tests have to be skipped, but if this works, then this will be easier than I thought.

I think it should be like that, we don't use the pass at all at the moment. 
but things are good to some extent, first I tried the tests with porting `simplifier.cpp` only without commenting any visitors or replacers and the failing tests were about 30, which is good.

File: integration_tests/run_tests.py (hunk 5 above)
Comment (assem2002) [reply integration_tests/run_tests.py]: Resolved

File: integration_tests/CMakeLists.txt (hunk 4 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:139]: I see! Thanks for looking into this.

**Comment by certik** [comment]
I see, if it is not doing anything, then everything should pass of course.

Perfect, finish the PR, we'll merge it and then we'll start enabling things.

File: integration_tests/run_tests.py
Comment (kmr-srbh) [inline integration_tests/run_tests.py:1]: Yes, this was another way we could do it. I think this approach is better than the short form I suggested, as we now test with all the available backends.

**Review by kmr-srbh — APPROVED** [review]

**Review by certik — APPROVED** [review]
I think that this is good to be merged.

**Comment by kmr-srbh** [comment]
It seems test 1106 is taking a lot of time?

image

**Comment by certik** [comment]
Do you know which test it is? We'll have to investigate.

**Comment by kmr-srbh** [comment]
It is `generic_interface_function_call_of_function_call.f90`.

**Comment by Pranavchiku** [comment]
I commented it, let's merge this PR so that we can put in other patches.

**Comment by Pranavchiku** [comment]
I don't think`generic_interface_function_call_of_function_call.f90` is the one.

**Comment by gxyd** [comment]
@assem2002 , how do we know that some integration test took a long time? Could you point me to a CI job or check which took a long time to run?

**Comment by gxyd** [comment]
> #### Test LLVM->WASM 10

one of the above comments mentioned *Test LLVM -> WASM 10*, was it the cancelled *Test LLVM ->WASM10" i.e. https://github.com/lfortran/lfortran/actions/runs/11439868624/job/31824489685?pr=5095?

**Comment by Pranavchiku** [comment]
Yes Gaurav, a few integration tests in `Test LLVM->WASM 10` and `Test LLVM->WASM 16` takes long time.

**Comment by assem2002** [comment]
@gxyd this one took over 3 hours https://github.com/lfortran/lfortran/actions/runs/11433321531/job/31805118345.

**Comment by gxyd** [comment]
Yes, I'm looking at it locally to find the culprit here.

**Comment by gxyd** [comment]
`implied_do_loops4` is one of the culprit here.

**Comment by assem2002** [comment]
Okay cool can you just push a commit with `NO-EXPERIMENTAL-SIMPLIFIER` option on that test?

**Comment by gxyd** [comment]
> Okay cool can youpush a commit with `NO-EXPERIMENTAL-SIMPLIFIER` option on that test?

Yes, let me identify the other culprits here as well.

File: integration_tests/CMakeLists.txt [hunk 6]
```diff
@@ -561,9 +568,9 @@ RUN(NAME format_11 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME format_12 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME format_13 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME format_14 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
-RUN(NAME format_15 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
-RUN(NAME format_16 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
-RUN(NAME format_17 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+RUN(NAME format_15 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc NO_EXPERIMENTAL_SIMPLIFIER)
```
Comment (gxyd) [inline integration_tests/CMakeLists.txt:571]: format_15 and format_16 raise an error for me locally on macOS m1, but I think they might pass on Linux (but I'm not sure).
```console
Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x12e2e - <unknown>!strlen
           1:  0x7c5 - <unknown>!append_to_string
           2: 0x9b4e - <unknown>!_lcompilers_string_format_fortran
           3:  0x38d - <unknown>!main
           4:  0x3e7 - <unknown>!_start
           5: 0x13e4c - <unknown>!_start.command_export
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information
    2: memory fault at wasm address 0x30312000 in linear memory of size 0x10000000
    3: wasm trap: out of bounds memory access

2/4 Test #1: format_15_simplifier .............***Failed    0.02 sec
Runtime Error : Got argument of type (CHARACTER), while the format specifier is (I)

3/4 Test #4: format_18_simplifier .............   Passed    0.02 sec
4/4 Test #3: format_17 ........................   Passed    0.02 sec

50% tests passed, 2 tests failed out of 4

Label Time Summary:
gfortran          =   0.10 sec*proc (4 tests)
llvm              =   0.10 sec*proc (4 tests)
llvm_wasm         =   0.10 sec*proc (4 tests)
llvm_wasm_emcc    =   0.10 sec*proc (4 tests)

Total Test time (real) =   0.03 sec

The following tests FAILED:
	  1 - format_15_simplifier (Failed)
	  2 - format_16_simplifier (Failed)
Errors while running CTest
Command failed.

```

Comment (assem2002) [reply integration_tests/CMakeLists.txt:571]: they fail for me too.

Comment (gxyd) [reply integration_tests/CMakeLists.txt:571]: so, it's possible that if we remove `NO_EXPERIMENTAL_SIMPLIFIER` here, the CI might still pass.

Comment (gxyd) [reply integration_tests/CMakeLists.txt:571]: Ah yes, I see here in the CI job: https://github.com/lfortran/lfortran/actions/runs/11433321531/job/31805118345, that those tests actually failed. Thanks @assem2002

**Comment by gxyd** [comment]
Let's see if the tests pass now, below were the files which needed `NO_EXPERIMENTAL_SIMPLIFIER`:
```console
format_15 (test fails)
format_16 (test fails)
implied_do_loops4 (takes too long)
format_17 (takes too long)
```

**Comment by assem2002** [comment]
Finally this PR passes. Thank you guys for the great help

**Comment by Pranavchiku** [comment]
Is it good to merge?

File: integration_tests/CMakeLists.txt [hunk 7]
```diff
@@ -1806,9 +1813,9 @@ RUN(NAME struct_type_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
...
-    EXTRA_ARGS --realloc-lhs)
+    EXTRA_ARGS --realloc-lhs NO_EXPERIMENTAL_SIMPLIFIER)
 RUN(NAME generic_interface_function_call_of_function_call LABELS gfortran llvm llvm_wasm llvm_wasm_emcc
-    EXTRA_ARGS --realloc-lhs)
-RUN(NAME elemental_function_overloaded_compare LABELS gfortran llvm llvm_wasm llvm_wasm_emcc
-    EXTRA_ARGS --realloc-lhs)
+    EXTRA_ARGS --realloc-lhs NO_EXPERIMENTAL_SIMPLIFIER)
 RUN(NAME elemental_function_scalar_array_arg LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+# RUN(NAME elemental_function_overloaded_compare LABELS gfortran llvm llvm_wasm llvm_wasm_emcc
+#     EXTRA_ARGS --realloc-lhs)
```
[the remark this answers is not in the document]
Comment (assem2002) [reply integration_tests/CMakeLists.txt:1821]: This appears to be causing no issue we uncomment it back.

**Comment by assem2002** [comment]
what I can think of is just the passes new order.

File: src/libasr/pass/pass_manager.h [hunk 8]
```diff
@@ -207,6 +210,83 @@ namespace LCompilers {
...
             c_skip_pass{false} {
+            _passes_with_experimental_simplifier = {
+                "global_stmts",
+                "init_expr",// This pass shouldn't be needed.
+                "function_call_in_declaration",
+                "implied_do_loops", // Should be implemented when optimisations for ImpliedDoLoop are possible in LFortran, until then not needed.
+                "openmp",
+                "simplifier", /* Verification checks to be implemented in this pass - 1. No array, user defined type variable should have a symbolic value. 2. Print, SubroutineCall, FileWrite, IntrinsicImpureSubroutine nodes shouldn't have non-Var arguments. 3. All expressions which need a temporary should be directly linked to a target via an assignment. 4. Sizes of auxiliary allocatables should be calculated using only Var nodes (with non-array symbols), or FunctionCall returning scalars. */
+                "nested_vars",
+                "transform_optional_argument_functions",
```
Comment (assem2002) [inline src/libasr/pass/pass_manager.h:221]: here

**Comment by certik** [comment]
Ok. I just merged it. Please submit a PR to enable that one test that was disabled.

**Pull request merged**